### ISSUE TITLE
Size & position not maintained on scaled displays

### ISSUE THREAD

**Issue by ronys** [pwsafe/pwsafe#1092]
On systems with high resolution displays and Scale > 100%, the last size & position of the main window is incorrectly restored.

**Comment by ronys** [comment]
Fixed in commit 1588ace06

**Comment by gillius** [comment]
This might have broken things for me, when upgrading from 3.64.1 to 3.65.1 the display appears as "blurry" upscaled again. This is something that has changed multiple times over the course of pwsafe (maybe last time it worked is after #966). I'm using 64-bit version of pwsafe and I confirmed in compatibility settings there is nothing checked in high DPI compatibility.

The initial dialog where you select a database is properly scaled with full resolution. Title bars on all windows are also fine. But once the application starts, the icons, menu bar, password tree, and edit entry are all upscaled and blurry, but the same "physical" size as the old 3.64.1.

I think it's awesome TOTP support came to 3.65. I have a separate TOTP app and maybe someday I will migrate the data over but for now as I'm not using that I downgraded back to 3.64.1 and it works. I did not try 3.65.0 but will if you feel it will help provide some context.

**Comment by gillius** [comment]
Actually, I just tested 3.65.0 64-bit version it does not work either, all the text is blurry. So something not listed in the release notes https://pwsafe.org/news.shtml probably affected something in 3.65.0 for me. I can file a separate issue if it makes sense.

Thanks for all of your efforts on maintaining this tool! For me, I'm OK with 3.64.1 for now this is not a hot item.

**Comment by ronys** [comment]
@gillius - what display resolution do you have, and what scale are you using?

A screenshot (of a test database) illustrating the problem would be great. Note that you have to uncheck "Exclude from screen capture" in Manage->Options->Security to take a screenshot.

**Comment by gillius** [comment]
@ronys thank you for the reply, I can open a separate issue if needed.

My resolution is 3840x2160. The scale is set to 150%. I am on Windows 11.

Here is version 3.65.01 64-bit about screen where you can see some properly scaled and other text not.
<img width="419" alt="v3 65 01 64-bit About" src="https://github.com/pwsafe/pwsafe/assets/1524332/8c6c3764-2609-41b8-91cb-e6fa8ce8ff67">
Same version with add entry screen where you can see it's rendered at lower resolution then scaled up in size
<img width="420" alt="v3 65 01 64-bit" src="https://github.com/pwsafe/pwsafe/assets/1524332/33affe98-f6ba-403a-aeea-36a1372fc5fb">

Old pwsafe v3.64.01 where the scaling is "native" at "full resolution" (i.e. the text is rendered at 150% instead of what appears to be rendered at 100% and raster upscaled)
<img width="332" alt="v3 64 01 64-bit" src="https://github.com/pwsafe/pwsafe/assets/1524332/ebf5ff09-c79a-4a1e-aa19-f6cd47f3ebf7">

**Comment by ronys** [comment]
@gillius , can you try the following:
Right click on the pwsafe.exe file in C:\Program Files\Password Safe, Properties->Compatibility->Change high DPI settings.

There, you can try changing Program DPI or High DPI scaling override.
Let me know if any setting makes a difference.

**Comment by gillius** [comment]
@ronys Program DPI option did not change the behavior for me, but overriding the scaling to application or system made it rendered at full resolution. System (enhanced) made it look just a little blurry but it seemed not as bad as with no configuration. However, in these modes at least with application scaling, every time opening and closing the database it shifted in +x and +y offset and width and height of the window increased. This is with 3.65.1 so it would have included fixes for this GH issue.

I didn't see any direct docs on how to build an executable, but a bunch of docs with choosing wxwidgets or not. Is this something reasonably buildable just by installing MSVC community + cmake and running cmake build? I'm a Java dev and haven't done C++ in almost 20 years but I might be willing to try out a git bisect if it's easy enough to build, although it could be up to a week before I can get to it, but I am not in a rush since I'm not using the new TOTP feature yet. I didn't see any commits that caught my eye as a likely cause though between the two versions.

**Comment by ronys** [comment]
@gillius Yes, cmake and MSVC community are what I'm using.
There are a couple of documents that are out of date, but should get your pointed in the right direction: README.CMAKE.md and README.WINDOWS.DEVELOPERS.html (note to self: unify and update).

As to commits, the following is probably what you're looking for:
1588ace06 [2 months ago] [GH1092] Fix restore size & pos on Scaled displays (bump rev to 3.65.1) [ronys]

**Comment by gillius** [comment]
@ronys I finally got to build it. I think there might be something in how it is built, because the versions I build from master branch (debug x64) and tag 3.65.1 (debug and release x64) work fine. I even tried to remove my `pwsafe.cfg` thinking maybe that has something to do with it, but didn't change anything. Here's a side by side screenshot of installed 3.65.1 next to git checkout of tag 3.65.1 running in debug mode (since if I try to run in release mode it prevents multiple-instances):

![image](https://github.com/pwsafe/pwsafe/assets/1524332/ad1e8981-0295-4f5c-8211-a3240ce3f1da)

The dialogs are also upscaled in github downloaded release file but not in locally built version:

![image](https://github.com/pwsafe/pwsafe/assets/1524332/1148454c-3db2-478b-a0d4-9c7b113b7ff9)

Since my build of 3.65.1 worked I didn't try to `git bisect` to find out where it "broke" between 3.64.1 and 3.65.0 (the GH download for 3.65.0 and 3.65.1 both get fuzzy upscaled for me).

I had some challenges building the project, but pieced it together from the 2 docs you provided. Neither were completely accurate for me.

### Install
1. Installed VS 2022 Community through winget: `winget install -e --id Microsoft.VisualStudio.2022.Community`
2. Installed cmake through winget: winget install -e --id Kitware.CMake
3. Launched VS 2022 IDE, skip account and open without code
4. Select Tools and Get Tools and Features
  1. In Workloads, select:
    1. Desktop development with C++
  1. In Individual Components search for and select:
    1. C++ CMake tools for Windows
    2. Windows 11 SDK
    3. C++ MFC for latest v143 build tools (x86 & x64)

**Notes:**
* If you skip CMake tools you'll get an error when running CMake that it cannot find any instance of visual studio
* If you skip Windows 11 SDK you'll get an error when running CMake "no cmake_cxx_compiler could be found"
* If you skip Desktop development with C++ or MFC, you'll be prompted to install them later when you open the generated solution

### Generate Solution
I used `cmake -S . -B _build` to generate the build files from the pwsafe folder, but the `cmake-gui` instructions from the README.CMAKE.md also work in this part I think.

I noticed during cmake configure it found my existing Python 3.11.9 installation. Is that required/useful? If it is maybe someone else following this would need Pythong. Also I didn't need to follow the directions to get gtest mentioned in README.CMAKE.md. It looked like it might have downloaded it during cmake configure?

Double click on the generated `_build/passwordsafe.sln` file to open Visual Studio.

Right click pwsafe project in the solution and pick "Set as startup project". Then pick "start without debugging" (ctrl-F5) or other desired option if you want debugging, change to Release build, etc. Once it starts, it will launch Password Safe. Note, it looks like in debug mode it uses default settings and closing app doesn't end the process but leaves it in system tray. I had to remember to exit it there before restarting app.

**Comment by gillius** [comment]
@ronys glad my experience could be useful.

Do you think it's the build OS, or the SDK installation? Can you build using latest Win 11 SDK in Win 10? What about Github Actions to build?

These are just thoughts, since I've not done a native project in a very long time, but such an issue even from modern Windows development seems surprising to me on this point. If there is anything else I can help with on this point let me know.

**Comment by ronys** [comment]
Hi @gillius,

I've updated the SDK on my Windows 10 machine, and would appreciate it if you could test the resulting build on your platform.
You can download the build artifacts from https://sourceforge.net/projects/passwordsafe/files/PreReleases/3.66.0pre/

Thanks!

**Comment by gillius** [comment]
@ronys looks like it's the same, I downloaded the pwsafe64-3.66.0pre from that folder and installed it.

![image](https://github.com/pwsafe/pwsafe/assets/1524332/787bfebd-a1c1-49f1-8428-41cdd0d58179)

As a test, I did a git checkout of the tag 3.66.0pre and did a build. This time I deleted my `_build` and followed the instructions from the readme where you said you can just open a cmake project with VS and built Windows x64 Debug pwsafe.exe and now I get the "blurry" version as well.

However, I run the `cmake -S . -B _build` and open the `.sln` from the `_build` folder and set pwsafe as startup project and pick Debug x64 and run it, and it looks fine:

![image](https://github.com/pwsafe/pwsafe/assets/1524332/43b171de-949e-4e95-85ca-e2569ad015be)

Just to make sure I wasn't crazy I re-opened the `pwsafe` folder directly, using VS CMake integration and it was still blurry:

![image](https://github.com/pwsafe/pwsafe/assets/1524332/b79c913a-c54a-445e-8d4d-dc89984800a9)

I've been out of VS for quite a long time so I'm a bit surprised it's different, but the UX is definitely different between the two methods and also the build process looks a lot different between the two in the sense that the output window's text output is totally different formats and makes it look like it's doing different things (although still a major part of both is compiling a bunch of files into obj which as expected). If you think any of that is useful to know I can bring it out.

**Comment by ronys** [comment]
So it looks like there are two variables: the build platform and the build system.
I've built the deliverables on Windows 11 using "Open folder", which configures cmake within the IDE and builds using Ninja,
and on Windows 10 by invoking cmake externally, and opening the created solution in the IDE.
These are in https://sourceforge.net/projects/passwordsafe/files/PreReleases/3.66.0pre/Win11Build/ and https://sourceforge.net/projects/passwordsafe/files/PreReleases/3.66.0pre/Win10SlnBuild/, respectively.

I'd greatly appreciate it if you can check these on your machine and let me know which, if any, give a clear text.
Thanks!

**Comment by gillius** [comment]
Note in both cases I was using the pwsafe64 exe version.

The W11Build version is not right:
![image](https://github.com/pwsafe/pwsafe/assets/1524332/4e518436-a972-4926-a15d-0bbe667070f9)

The Win10InSBuild version does work:
![image](https://github.com/pwsafe/pwsafe/assets/1524332/4a7be83d-a1bd-45f6-ab36-8e7891dc7ce6)

**Comment by gillius** [comment]
@ronys as I reported before the build you gave me the Win10InSBuild does have clear text, but I realize now that it might have broken this issue (or maybe it was broken all along?) -- now every time I start the app the window is bigger and top left corner is shifted down and right.

**Comment by arzu3649** [comment]
Hey guys. Is this issue really fixed?
Every time I start my 3.66.01 Password Safe on my 125% scaled monitor - it opens up shifted far left and down. It is annoying, as I have to move and resize it to get to the search field. Screenshot attached.
![image](https://github.com/user-attachments/assets/eb3ea4c9-1bd1-4766-b89d-e67686ad3630)

**Comment by gillius** [comment]
@arzu3649 I ended up reverting back to 3.64.01, which the windows moves slowly to the right on each restart but not anywhere as much and doesn't have the resizing issue. In my last comments, ronys built 2 different versions, and one of them fixed the text upscaling but the issue of the size and position was returned. I haven't follow up with it since. I think my build of 3.65 was working at one point and I'm not sure why I didn't decide to use it but after my last post in May I stopped working on the issue since I didn't need the main feature of the newer version which was TOTP.

**Comment by arzu3649** [comment]
@ronys - can we re-open this issue?

**Comment by ronys** [comment]
@arzu3649 - what version of Windows are you using, and what's your screen resolution?

**Comment by arzu3649** [comment]
Windows 10, 125% font scaling, 1920x1080 and this:
![image](https://github.com/user-attachments/assets/75152e74-e1b0-4678-8abf-007cb3c1701d)

**Comment by bb09961** [comment]
I also have this issue on the latest version.

Windows 10, 118% custom scaling,
2 monitors:  2560x1440  and 1920x1080

I've tried all of the High DPI scaling settings and they do not help.

The issue i see, is when the application is shut down, then re-started, the window will reset to a default size that runs off the bottom and right of the screen.  If the window would open on my larger screen, it may fit in on the screen, but it always opens on the smaller screen.   But most importantly, it doesn't remember the previous size and position of the application window.

**Comment by aeadie369** [comment]
Hi, I have the same issue for a long time. "On systems with high resolution displays and Scale > 100%, the last size & position of the main window is incorrectly restored", the same issue as arzu3649 above. Every time I start my 3.67 Password Safe on my 125% or 200% scaled monitor it opens up shifted far right and down on the main 4k monitor and the window is way taller than the screen. It is annoying, as I have to move and resize it to get to the search field. The scaling on the 4k monitor is as expected and sharp, the scaling on the 2k monitor is outsized (too large fonts) and some font sizes are not matching settings. I cant find any way to set the start window size in pwsafe.cfg or the helpdocs.

Windows 11 Pro 24H2 Build 26100.2894
1 x 4k monitor @ 200% Scaling
1 x 2k monitor @125% scaling

Based on this thread it might be easiest to allow the user optionally to specify a window size, it would save you as developer trying to find the best fit for all users across all scaling ;-)

**Comment by RenamedPlayer** [comment]
Windows 10, Version 3.67, one screen 2560x1440 Points, 125%.
Only when I don't change X, Y, Width, Height, no moving, no sizing, then the window is opened with the coordinates from closing. Changing Width/Height increases X/Y, changing X/Y increases Width/Height — up to outside off the screen.
Work-Around: Try to move it to somewhere top/left on the screen, with small Width and Height. Maybe it will then have useable coordinates. Dont change them.
Maybe 15, 20 Years ago someone explained the pointlessness of saving window coordinates in such this way with resolution, font size etc. That would be an endless story. Save them absolute, on load cut them off to the current screen with somewhat minimum Width and Height. Thats the best way.

**Comment by bb09961** [comment]
The workaround above worked for me.   I put the Password Safe window in the very top left corner of my screen and it has stayed in this location and retained its size after a number of application restarts and computer reboots.   It is not ideal but it will work until this bug can be fixed.

Thanks for the tip.

**Comment by ColeBantam** [comment]
Still not working for me in Version 3.72 :(
Window size and position still not saved.

Using 3840x2160 with 175% scaling
Windows 11 25H2
No options in "compatibility settings" checked.
Pwsafe.cfg deleted.

Any other things I need to delete, when coming from the last working version (3.64.1)?

PS: Fonts still blurry too ...
